client/state/purchases/index.js

```diff
@@ -0,0 +1,173 @@
+	PURCHASES_USER_FETCH,
+	PURCHASE_REMOVE_COMPLETED,
+	PURCHASES_SITE_FETCH_COMPLETED,
+	PURCHASES_USER_FETCH_COMPLETED,
+	PURCHASE_REMOVE_FAILED,
+	PURCHASES_SITE_FETCH_FAILED,
+	PURCHASES_USER_FETCH_FAILED,
+	PRIVACY_PROTECTION_CANCEL_COMPLETED,
+	PRIVACY_PROTECTION_CANCEL_FAILED
+} from 'state/action-types';
+
+/**
+ * Constants
+ */
+const initialState = {
+	data: [],
+	error: null,
+	isFetchingSitePurchases: false,
+	isFetchingUserPurchases: false,
+	hasLoadedSitePurchasesFromServer: false,
+	hasLoadedUserPurchasesFromServer: false
+};
+
+function updatePurchaseById( state, id, properties ) {
+	return assign( {}, state, {
+		data: state.data.map( purchase => {
+			if ( id === purchase.id ) {
+				return assign( {}, purchase, properties );
+			}
+			return purchase;
+		} )
+	} );
+}
+
+/**
+ * Overwrites the purchases in the store with the purchases from the new purchases array
+ * that share the same `id` value.
+ *
+ * @param {array} existingPurchases - an array of purchases in the store
+ * @param {array} newPurchases - an array of purchases fetched from the API
+ * @return {array} An array of purchases
+ */
+function overwriteExistingPurchases( existingPurchases, newPurchases ) {
+	let purchases = newPurchases;
+
+	existingPurchases.forEach( purchase => {
+		if ( ! find( purchases, { id: purchase.id } ) ) {
+			purchases = purchases.concat( purchase );
+		}
+	} );
+
+	return purchases;
+}
+
+/**
+ * Removes purchases that are missing from the new purchases array and match the given predicate.
+ *
+ * @param {array} existingPurchases - an array of purchases in the store
+ * @param {array} newPurchases - an array of purchases fetched from the API
+ * @param {object} predicate - the predicate to check before removing the item from the array.
+ * @return {array} An array of purchases
+ */
+function removeMissingPurchasesByPredicate( existingPurchases, newPurchases, predicate ) {
+	return existingPurchases.filter( purchase => {
+		if ( matches( predicate )( purchase ) && find( newPurchases, { id: purchase.id } ) ) {
+			// this purchase is present in the new array
+			return true;
+		}
+
+		if ( ! matches( predicate )( purchase ) ) {
+			// only overwrite remove purchases that match the predicate
+			return true;
+		}
+
+		// the purchase doesn't match the predicate or is missing from the array of new purchases
+		return false;
+	} );
+}
+
+function updatePurchases( existingPurchases, action ) {
+	let purchases, predicate;
+
+	if ( PURCHASES_SITE_FETCH_COMPLETED === action.type ) {
+		predicate = { siteId: action.siteId };
+	}
+
+	if ( PURCHASES_USER_FETCH_COMPLETED === action.type ||
+		PURCHASE_REMOVE_COMPLETED === action.type ) {
+		predicate = { userId: action.userId };
+	}
+
+	purchases = removeMissingPurchasesByPredicate( existingPurchases, action.purchases, predicate );
+	purchases = overwriteExistingPurchases( purchases, action.purchases );
+
+	return purchases;
+}
+
+export const items = ( state = initialState, action ) => {
```

Comment:
I used `items` because it seems to be a relatively common pattern in the state tree, e.g. [state.posts.items](https://github.com/Automattic/wp-calypso/blob/master/client/state/sites/reducer.js#L101) and [state.sites.items](https://github.com/Automattic/wp-calypso/blob/master/client/state/sites/reducer.js#L101).